controller/test/src/ArduinoSim.cpp

```diff
@@ -0,0 +1,100 @@
+/* Copyright 2020, RespiraWorks
+
+  This file is part of VentilatorSoftware.
+
+  VentilatorSoftware is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  VentilatorSoftware is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with VentilatorSoftware.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+/*
+ module contributors: verityRF
+ The purpose of this module is to recreate fake versions of Arduino routines so that unit tests of the controller modules can be done on a x86 host.
+ This module should never end up on an actual Arduino target.
+*/
+
+//@TODO: Write unit tests for this module
+
+#include "ArduinoSim.h"
+#include <stdlib.h>
+#include <math.h>
+
+const int ANALOG_SIGNAL_DEPTH = 256; //how (maximally) long are the simulated analog signals
+const int NUM_ANALOG_CHANNELS = 4; //how many channels (analog pins) are supported
+const float ADC_VREFP = 5.0f; //Arduino ADC VREF_P setting/value [V]
+const float ADC_COUNTS_PER_VOLT = 1024.0f / ADC_VREFP; //10 bit ADC with given ADC VREF_P [Count/V]
+
+static float analogSignalMemory[NUM_ANALOG_CHANNELS][ANALOG_SIGNAL_DEPTH];
+static int analogSignalMemoryIndexes[NUM_ANALOG_CHANNELS];
+
+void delay(unsigned long count) {
+	//method to remain unimplemented; isn't useful from a unit test perspective
+}
+
+int getSignalChannelDepth() {
+	return ANALOG_SIGNAL_DEPTH;
+}
+
+void resetAnalogMemoryIndexes() {
+	for (int i = 0; i < NUM_ANALOG_CHANNELS; i++) {
+		analogSignalMemoryIndexes[i] = 0;
+	}
+}
+
+void createDynamicAnalogSignal(int pin, float *buffer, int count) {
+	if (buffer == NULL) {
+		throw "Input buffer pointer is null";
```

Comment:
Perhaps we should use assert() rather than throw?  When compiling for the controller obviously we'll have exceptions disabled (if not we should do that!), so presumably none of our code will be exception-safe anyway.

Reply:
ArduinoSim does not end up on the Arduino target. It is run on x86 hosts only. Thus there is no concern for assert/throw relative to it being on the Arduino.

Reply:
Understood.

What I guess I am saying is, because none of the Arduino code is exception-safe, `throw` is equivalent to `assert(false && "foo")`, right?  If so, I wonder if we should use assert() so that we do not believe that e.g. it would be safe to try to catch an exception within a testing framework and continue running.

Reply:
I did figure we'd want to change how we handle errors in this module relative to the overall testing framework. I have no problem going one way or the other, wanted some input from the folks who were more focused on the testing.

Reply:
I'll refactor to asserts instead of throws and make sure it works with platformIO properly as a separate change.